site/cds_rdm/tasks.py

```diff
@@ -140,3 +147,430 @@ def sync_users():
+            )
+            updated = True
+        return updated
+
+    def _check_if_update_needed(user, name, is_orcid=False, deprecate=False):
+        """Check if the name needs to be updated.
+
+        param user: The user object.
+        param name: The name dictionary.
+        param is_orcid: If the name passed is an ORCID value.
+        param deprecate: If the name should be marked as unlisted.
+        """
+
+        updated = False
+        updated_name = {**name}
+
+        # We only update the names if it's a non ORCID value, for ORCID values
+        # we rely on the ORCID harvester to update the given name, family name
+        if not is_orcid:
+            if user.user_profile.get("given_name") != name.get("given_name"):
+                updated_name["given_name"] = user.user_profile.get("given_name")
+                updated = True
+
+            if user.user_profile.get("family_name") != name.get("family_name"):
+                updated_name["family_name"] = user.user_profile.get("family_name")
+                updated = True
+
+        update_functions = [
+            _add_affiliations,
+            _add_orcid,
+            _add_person_id,
+            _add_or_update_props,
+            _add_or_update_email,
+        ]
+
+        for update_func in update_functions:
+            updated = update_func(user, name, updated_name, updated=updated) or updated
+
+        if deprecate and "unlisted" not in updated_name.get("tags", []):
+            if updated_name.get("tags"):
+                updated_name["tags"].append("unlisted")
+            else:
+                updated_name["tags"] = ["unlisted"]
+            updated = True
+
+        return updated_name if updated else None
+
+    def _update_name(user, name_dict, is_orcid=False, deprecate=False, uow=None):
+        """Updates the name with the new values.
+
+        param user: The user object.
+        param name_dict: The name dictionary.
+        param is_orcid: If the name passed is an ORCID value.
+        """
+        updated_name = _check_if_update_needed(user, name_dict, is_orcid, deprecate)
+        if updated_name:
+            try:
+                names_service.update(
+                    system_identity, name_dict.get("id"), updated_name, uow=uow
+                )
+            except ValidationError as e:
+                current_app.logger.error(f"Error updating name for user {user.id}: {e}")
+
+    def _create_new_name(user, deprecate=False, uow=None):
+        """Creates a new name for the user.
+
+        param user: The user object.
+        """
+
+        default_props = _get_default_props(user.id)
+        name = {
+            "id": str(user.user_profile.get("person_id")),
+            "props": default_props,
+            "identifiers": [
+                {"scheme": "cds", "identifier": str(user.user_profile.get("person_id"))}
+            ],
+        }
+
+        if user.user_profile.get("given_name"):
```

Comment:
Maybe use yet-another utiliy here?
`add_if_exists(name, "given_name", user_profile.get("given_name"))`?

Reply:
Hmm we could definitely do it but, it would be a complex one if we want to cover all the cases as we do not always assing the value directly, sometimes we build a dict as for the affiliations, for others like affiliations we want to append and for the props they are nested. I thought it would simpler like this but I am happy to update it if required.